### PR TITLE
Add general entropy functionality to EOS

### PR DESCRIPTION
**Pull request by jhp-lanl**
## PR Summary

This MR adds the basic entropy hookups for entropy lookup functions to be exposed in the general EOS type. 

There are five main changes in this effort:

1. Add entropy lookups to the `Variant` class. This basically forces the EOS API to now provide entropy lookups so it needs to be implemented in some form everywhere.
2. Add a default function to raise an error that entropy isn't implemented and have the entropy functions call this function by default.
3. Use the ideal gas EOS as a prototype for adding entropy to other EOS. The functional form is fairly
4. Incorporate entropy into the modifiers correctly. I could have just used the error functionality, but it was actually fairly easy since most modifiers won't modify the entropy. I had to prove to myself how entropy should be modified with scaling ratios, but I think it's the correct approach.
5. Add a lot of documentation about entropy and complete vs incomplete EOS.

I wanted to implement entropy for at least one analytic EOS in this MR because I think it exposes important implications for other analytic EOS. Specifically, it demonstrates that the entropy is divergent at zero temperature for EOS with constant heat capacity, which further emphasizes the importance of the reference state in the entropy calculation.

There are a _lot_ of changes to the documentation where I wanted to add what I've learned in researching how to implement entropy. There are also some cosmetic changes that can be ignored (especially in the Vinet EOS) that just
made things easier for me to read in text form.

To do:
- [x] Create a test for the ideal gas entropy

First step on addressing https://github.com/lanl/singularity-eos/issues/210

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- N/A Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file
- N/A If preparing for a new release, update the version in cmake.

### CODE REVIEW

**Comment by jhp-lanl** [comment]
@aematts can you review this?

Also @Yurlungur and @dholladay00 can you give things a brief look to make sure you agree with my strategy for defaulting EOS to throw errors?

**Comment by jhp-lanl** [comment]
I should add that I haven't built the code yet so there's probably lots of little issues I need to fix tomorrow. I just wanted to get this up so people could start looking at things.

File: doc/sphinx/src/models.rst
Comment (Yurlungur) [inline doc/sphinx/src/models.rst:250]: Maybe make this a warning instead of a note?

File: singularity-eos/eos/eos_base.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_base.hpp:300]: If the compiler is still yelling at us for this decorator we could remove the `[[noreturn]]` from `portable_errors`... Not worth dealing with here but food for thought

File: singularity-eos/eos/eos_davis.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_davis.hpp:72]: I like the explicit stubbing out here. :+1:

**Review by Yurlungur — COMMENTED** [review]
:+1: This all seems reasonable to me.

File: singularity-eos/eos/eos_eospac.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp:284]: Absolutely! EOSPAC, spiner, and Davis should all get entropy. I just figured this MR could use an ideal gas as a test case (I'm planning to add tests). I am breaking it up into two MRs for simplicity.

File: singularity-eos/eos/eos_davis.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_davis.hpp:72]: Yeah I thought about trying to do it implicitly, but a) I couldn't figure out a way to do that from the base class while also providing the vector overloads, and b) I figured that it's probably better for every EOS to have the interface entropy functions explicitly defined so that somebody creating a new EOS will understand what to do better.

Comment (Yurlungur) [reply singularity-eos/eos/eos_davis.hpp:72]: Yep I think that makes sense. :+1:

File: singularity-eos/eos/eos_eospac.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_eospac.hpp:284]: Ah cool yeah that sounds good.

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [reply doc/sphinx/src/models.rst:250]: I had been reserving the warnings for EOS that just plain don't provide entropy, which is why I chose notes here. My thinking was that for EOS where entropy isn't enabled, your code will crash if you try to access it, so that's more serious.

Comment (Yurlungur) [reply doc/sphinx/src/models.rst:250]: Sure that's fair.

File: singularity-eos/eos/eos_base.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp]: Hmmmm yeah maybe we do want to do this since it will only truly not return when the code aborts. When we throw an exception, the function should still return, right?

I'll change this.

**Comment by jhp-lanl** [comment]
This MR depends on https://github.com/lanl/singularity-eos/pull/234

**Comment by jhp-lanl** [comment]
@Yurlungur I _think_ this is ready for friendly testing. I hooked it up to a basic test and squashed the resulting bugs. I'll commit the test tomorrow when I can expand the coverage.

**Comment by jhp-lanl** [comment]
@aematts any thoughts?

@dholladay00 I don't think there's much for you to review, but make sure you agree with how I'm implementing the default entropy error throwing

**Comment by jhp-lanl** [comment]
> @aematts any thoughts?

I should add, I'd greatly appreciate your eyes on the documentation, @aematts . I've written some stuff up about EOS theory that mirrors the document you wrote (which I cited). I'd love your feedback there if you can give it.

File: doc/sphinx/src/models.rst
Comment (Yurlungur) [inline doc/sphinx/src/models.rst:211]: I don't know if this is true. I think we let the user specify an initial guess don't we?

File: singularity-eos/eos/eos_base.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_base.hpp:395]: style nitpick:
```suggestion
    std::strcat(msg, eosname);
    std::strcat(msg, "' EOS");
```

File: singularity-eos/eos/eos_base.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_base.hpp:392]: Good solution :+1:

**Review by Yurlungur — APPROVED** [review]
This looks good. Thanks for doing this, @jhp-lanl ! A high priority should be enabling entropy in the tabulated EOS's but getting the framework in place is an important first step.

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [reply doc/sphinx/src/models.rst:211]: Here is my quick survey of the `DensityEnergyFromPressureTemperature()` function in our EOS. It looks like JWL is the only analytic that allows for a user-supplied guess in the root find while all others behave as I describe or have an analytic inversion.

- Ideal gas: analytic inversion
- Gruneisen: analytic inversion for energy. Density is analytically inverted in expansion but uses a root find in compression with an ambient density guess
- JWL: Root find with user-supplied guess through the function argument if it's large enough (does not use lambda)
- Davis Reactants: Root find with ambient density guess (no user input)
- Davis Products: Root find with ambient density guess (no user input)
- Vinet: Not implemented error
- EOSPAC: Whatever EOSPAC does under the hood
- Spiner: Can utilize `lambda` for a user-supplied guess, but otherwise ignores the density function argument

I'll note that spiner can take user-supplied guesses in the lambda parameter but I'll keep the language about analytic EOS since I think it broadly applies.

Comment (jhp-lanl) [reply doc/sphinx/src/models.rst]: I'll create an issue to possibly unify this behavior in the base class and provide custom implementations where appropriate.

Comment (jhp-lanl) [reply doc/sphinx/src/models.rst]: Ah never mind, we already have an issue. I'll rename it
https://github.com/lanl/singularity-eos/issues/199

**Comment by aematts** [comment]
> > @aematts any thoughts?
> 
> I should add, I'd greatly appreciate your eyes on the documentation, @aematts . I've written some stuff up about EOS theory that mirrors the document you wrote (which I cited). I'd love your feedback there if you can give it.

I have been to a conference this week and have had no time to even read my emails. I will get to this sometime next week.

**Comment by jhp-lanl** [comment]
> > > @aematts any thoughts?
> > 
> > 
> > I should add, I'd greatly appreciate your eyes on the documentation, @aematts . I've written some stuff up about EOS theory that mirrors the document you wrote (which I cited). I'd love your feedback there if you can give it.
> 
> I have been to a conference this week and have had no time to even read my emails. I will get to this sometime next week.

I'm going to delay until the end of the day **Monday** to merge this, but even if you don't have time to make comments by then, please feel free to make comments even after these changes are merged. It will then be fairly easy to just create a new PR to address your comments on these changes, especially if most of the comments are in the documentation.

Normally I'd just wait for all the feedback to come in, but given that this MR has been open for a while already without feedback and that Jonah has a student waiting for these changes, I figure it's possible to merge these changes with the expectation that more changes will be required down the road. Besides, I'll need to create another MR to add in EOSPAC, spiner, and Davis entropy capabilities.

**Comment by dholladay00** [comment]
@jhp-lanl what's the timeline for merging this?

**Comment by jhp-lanl** [comment]
@dholladay00 sorry last week was pretty crazy. I was going to merge it last week but forgot about it. I'll try to resolve the conflicts in the changelog this afternoon and get it merged.

@aematts feel free to give feedback after the MR has been merged and I can address your feedback in subsequent MRs

**Comment by jhp-lanl** [comment]
> I'll try to resolve the conflicts in the changelog this afternoon and get it merged.

I'll make this a top priority for tomorrow

**Pull request merged**
